src/test/groovy/com/calgaryscientific/gradle/VeracodeTaskTest.groovy

```diff
@@ -192,16 +194,69 @@ class VeracodeTaskTest extends Specification {
 
         when:
         project.plugins.apply('com.calgaryscientific.gradle.veracode')
-        project.ext.veracodeSetup = vs
+        project.veracodeSetup.filesToUpload = vs.filesToUpload
 
         then:
         VeracodeSetup vsRead = project.findProperty("veracodeSetup") as VeracodeSetup
         Set<File> expected = [buildFile] as Set
         assert vsRead.filesToUpload == expected
-        def task = project.tasks.getByName("veracodeUploadFile")
+        VeracodeUploadFileTask task = project.tasks.getByName("veracodeUploadFile") as VeracodeUploadFileTask
+        assert task.veracodeSetup.filesToUpload == vsRead.filesToUpload
         assert task.getFileSet() == expected
         def _ = vsRead.filesToUpload.add(buildFile)
         assert vsRead.filesToUpload == [buildFile, buildFile] as Set
     }
 
+    def 'Test VeracodeUploadFile Task'() {
+        given:
+        VeracodeAPI veracodeAPIMock = Mock(VeracodeAPI, constructorArgs: ["", "", null, null])
+        String xmlStr = '''<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<filelist xmlns="something" xmlns:xsi="something" filelist_version="1.1">
+    <file file_id="1" file_md5="d98b6f5ccfce3799e9b60b5d78cc1" file_name="file1" file_status="Uploaded"/>
+    <file file_id="2" file_md5="68a7d8468ca51bc46d5b72d485022" file_name="file2" file_status="Uploaded"/>
+    <file file_id="3" file_md5="2459464ff4bf78dd6f09695069b52" file_name="file3" file_status="Uploaded"/>
+</filelist>
+'''
+        Project project = new ProjectBuilder().build()
+        project.plugins.apply('com.calgaryscientific.gradle.veracode')
+        VeracodeSetup vs = new VeracodeSetup()
+        vs.filesToUpload = project.fileTree(dir: testProjectDir.root, include: '**/*').getFiles()
+        project.veracodeSetup.filesToUpload = vs.filesToUpload
+        VeracodeUploadFileTask task = project.tasks.getByName("veracodeUploadFile") as VeracodeUploadFileTask
+        task.veracodeAPI = veracodeAPIMock
+
+        when:
+        task.run()
+
+        then:
+        1 * veracodeAPIMock.uploadFile(_, _) >> {
+            return xmlStr
+        }
+    }
+
+    def 'Test VeracodeUploadFile Task failure'() {
+        given:
+        VeracodeAPI veracodeAPIMock = Mock(VeracodeAPI, constructorArgs: ["", "", null, null])
+        String xmlStr = '''<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<error>Could not upload file</error>
+'''
+        Project project = new ProjectBuilder().build()
+        project.plugins.apply('com.calgaryscientific.gradle.veracode')
+        VeracodeSetup vs = new VeracodeSetup()
+        vs.filesToUpload = project.fileTree(dir: testProjectDir.root, include: '**/*').getFiles()
+        project.veracodeSetup.filesToUpload = vs.filesToUpload
+        VeracodeUploadFileTask task = project.tasks.getByName("veracodeUploadFile") as VeracodeUploadFileTask
+        task.veracodeAPI = veracodeAPIMock
+        task.waitTimeBetweenAttempts = "0"
+
+        when:
+        task.run()
```

Comment:
can we rename run to something to indicate that the task is uploading artifacts to veracode to be analyzed?

Reply:
no, `run` is the method that implements the task interface.

Reply:
I see, the upload task extends from abstract class (contains abstract run function), so the upload task need to override the abstract run method.

I guess, what's the reasons for having VeraCodeTask abstract class?

Reply:
My bad, I didn't look before answering. 

> what's the reasons for having VeraCodeTask abstract class?
At the very least to have the common required and optional argument handling code.